### ISSUE TITLE
SpaceInsideHashLiteralBraces fails on nested hashes

### ISSUE THREAD

**Issue by brewster1134** [bbatsov/rubocop#2595]
i noticed that a hash like `h = { a: 1, { b: 2 }}` will fail since it expects `h = { a: 1, { b: 2 } }`
is this desired/standard style? i have always written it as the former. if the former should be allowed, i could take a stab at writing a spec for it (i am not familiar with this codebase however ;)

**Comment by alexdowad** [comment]
So you are suggesting that there should be space inside a hash literal... unless the last value (or first key) is itself a hash?

**Comment by brewster1134** [comment]
exactly.  i adopted that style after seeing it used in really popular codebases (rails & mongoid off the top of my head). e.g. the following hashes would be valid

```
{{ a: 1 }, b: 2 }
{ a: 1, {b: 2 }}
{ a: 1, { b: 2 }, c: 3 }
```

**Comment by jawshooah** [comment]
Guessing that the second should be

```
{ a: 1, { b: 2 }}
```

**Comment by brewster1134** [comment]
@jawshooah yup, updated it! thnx!

**Comment by brewster1134** [comment]
i have also found instances from rspec where there is a hash argument in a Proc

```
# meh
subject { Foo.new({ foo: 'bar' }, type: :hash) }

# looks better (to me)
subject{ Foo.new({ foo: 'bar' }, type: :hash )}
```

lots of edge cases with this cop it seems...

**Comment by jonas054** [comment]
```ruby
subject{ Foo.new({ foo: 'bar' }, type: :hash )}
```
Examples don't give the whole picture. It better if you can express all rules that would define such a style. Something like:
* The opening block brace should be preceded by a space, except in the following conditions:
  * ...
* There should be space inside bock braces except for:
  * ...
* There should be space inside hash literal braces except for:
  * ...
* There should be no space inside method call parentheses except for:
  * ...

There's a lot to be said for simple rules. It would be interesting to see these rules listed, but I'm not too keen on making things so complicated.

**Comment by alexdowad** [comment]
@brewster1134 What do you think about cases where a hash literal appears inside a block, like this:

```ruby
method { |h| h.merge { key: val }}
```

Do you prefer to see a space between the concluding braces, or not?

**Comment by brewster1134** [comment]
@alexdowad the convention i have seen, would be just like how you have it. i think basically the rule is you always want to avoid `) }`, `] }`, `} }`

@jonas054 when i have some time, i will sit down and define some specific rules _(unless the examples above cover everything ha!)_.  I just need to sit down and think it out... soon.... :smile:

**Comment by brewster1134** [comment]
@alexdowad so after sitting down and writing out a variety of use cases, i think it should actually only apply to braces.

* when 2 or more opening braces are separated with a space(s), the spaces should be removed, and a single space should follow the last brace: `{ {{ {` -> `{{{{ `
* when 2 or more closing braces are separated with a space(s), the spaces should be removed, and a single space should proceed the first brace: `} }} }` -> ` }}}}`

**Comment by alexdowad** [comment]
Can you suggest a name for this name style? We already have `space` and `no_space`.

**Comment by brewster1134** [comment]
`smart_spacing`? `compact_spacing`? umm `weighted_spacing`?
im not great with naming :smile: it seems like name should also highlight the fact the space is on the left or right depending on if its an opening/closing brace _(not sure if that was clear from the bulleted examples above)_
